File: client/dots.py
from typing import Tuple, Set, FrozenSet, Optional

from PyQt5.QtGui import QPainter, QPen, QColor, QBrush
from PyQt5.QtCore import QPoint, QSize, Qt


class DotsSettings:
	def __init__(self):
		self.colors = {
			0: Qt.red,
			1: Qt.blue,
			2: Qt.green,
			3: Qt.yellow
		}
		self.nicks = {
			0: 'red',
			1: 'blue',
			2: 'green',
			3: 'yellow'
		}


class DotsManager:
	empty = -1
	empty_eaten = -2
	
	def __init__(self, players_number: int):
		if not 0 < players_number < 5:
			raise AttributeError(f"Only from 1 to 4 player can play dots, not {players_number}.")
		self.players_number = players_number
		self.settings = DotsSettings()
	
	def _getPlayer(self, player: int) -> int:
		if not 0 <= player < self.players_number:
			raise AttributeError(f"It is only {self.players_number} players in {self}, player {player} does not exist.")
		return player
	
	def real(self, player: int) -> int:
		return self._getPlayer(player)
	
	def isEaten(self, dot: int) -> bool:
		return self.eaten(self.player(dot)) == dot

	def eaten(self, player: int) -> int:
		if player == self.empty:
			return self.empty_eaten
		return self._getPlayer(player) + self.players_number
	
	def color(self, dot: int) -> QColor:
		if dot in {self.empty, self.empty_eaten}:
			return Qt.transparent
		return self.settings.colors[self.player(dot)]
	
	def player(self, dot: int) -> int:
		if dot in {self.empty, self.empty_eaten}:
			return -1
		return dot % self.players_number


class Chunk:
	def __init__(self, x: int, y: int, dots_manager: DotsManager):
		self.x = x
		self.y = y
		self.dots_manager = dots_manager
		self.map = [[dots_manager.empty for _ in range(16)] for _ in range(16)]
		self.tracks: Set[FrozenSet[Tuple[int, int], Tuple[int, int]]] = set()
		self.eater = [[-1 for _ in range(16)] for _ in range(16)]
	
	def isEmpty(self):
		return self.map == [[self.dots_manager.empty for _ in range(16)] for _ in range(16)]
	
	def drawLines(self, painter: QPainter, tx: int, ty: int) -> None:
		line_pen = QPen(Qt.black)
		line_pen.setWidth(1)
		painter.setPen(line_pen)
		for i in range(16):
			painter.drawLine(self.x * 16 + tx, (self.y + i) * 16 + 8 + ty,  # \ horizontal line
			                 self.x * 16 + 256 + tx, (self.y + i) * 16 + 8 + ty)  # /
			painter.drawLine((self.x + i) * 16 + 8 + tx, self.y * 16 + ty,  # \ vertical line
			                 (self.x + i) * 16 + 8 + tx, self.y * 16 + 256 + ty)  # /
	
	def drawDots(self, painter: QPainter, tx: int, ty: int, dots: 'Dots') -> None:
		for track in self.tracks:
			track = tuple(track)
			max_dot = max(track)
			painter.setPen(self.dots_manager.color(self.map[max_dot[0] % 16][max_dot[1] % 16]))
			painter.drawLine(track[0][0] * 16 + 8 + tx, track[0][1] * 16 + 8 + ty,
	                 track[1][0] * 16 + 8 + tx, track[1][1] * 16 + 8 + ty)

		painter.setPen(Qt.transparent)
		for x in range(16):
			for y in range(16):
				painter.setBrush(
					self.dots_manager.color(self.map[x][y])
				)
				painter.drawEllipse((self.x + x) * 16 + 5 + tx, (self.y + y) * 16 + 5 + ty, 6, 6)
				if not self.dots_manager.isEaten(self.map[x][y]):
					continue
				fill_brush = QBrush(self.dots_manager.color(self.eater[x][y]))
				fill_brush.setStyle(Qt.BDiagPattern)
				painter.setBrush(fill_brush)
				before = {(0, -1): (-1, -1),
				          (1, 0): (1, -1),
				          (0, 1): (1, 1),
				          (-1, 0): (-1, 1)
				          }
				poly = [self.dotCoord(xx, yy) + QPoint(tx, ty)
				        for xx, yy in [(x, y - 1), (x + 1, y), (x, y + 1), (x - 1, y)]]
				for val in before:
					if dots.getDot(self.x + x + before[val][0], self.y + y + before[val][1]) == self.eater[x][y]:
						poly.insert(poly.index(self.dotCoord(val[0] + x, val[1] + y) + QPoint(tx, ty)),
						            self.dotCoord(before[val][0] + x, before[val][1] + y) + QPoint(tx, ty))
				painter.drawPolygon(*poly)
	
	def dotCoord(self, x: int, y: int) -> QPoint:
		return QPoint((self.x + x) * 16 + 8, (self.y + y) * 16 + 8)


class Dots:
	def __init__(self, players_number: int = 2):
		self.dots_manager = DotsManager(players_number)
		self.chunks = [[Chunk(x * 16, y * 16, self.dots_manager) for y in range(16)] for x in range(16)]
		self.tracks: Set[FrozenSet[Tuple[int, int], Tuple[int, int]]] = set()
		self.turning_player = 0
		self.cam_x: int = 2048
		self.cam_y: int = 2048
		self.scale: float = 2
		self.counters = [0] * players_number

	def addTrack(self, track: Tuple[Tuple[int, int], ...]) -> None:
		for i in range(len(track) - 1):
			max_dot = max(track[i], track[i + 1])
			self.chunks[max_dot[0] // 16][max_dot[1] // 16].tracks.add(frozenset((track[i], track[i + 1])))

	def changeScale(self, delta: float, size: QSize) -> None:
		self.scale += delta
		self.scale = max(self.scale, 2)
		self.scale = min(self.scale, 6)
		self.translate(QPoint(), size)
	
	@staticmethod
	def circle(dx: int, dy: int, radius: int) -> Set[Tuple[int, int]]:
		answer = set()
		for x in range(-radius, radius + 1):
			for y in (radius - abs(x), -radius + abs(x)):
				answer.add((x + dx, y + dy))
		return answer

	def draw(self, size: QSize, painter: QPainter, cursor: Optional[QPoint] = None) -> None:
		painter.save()
		painter.scale(self.scale, self.scale)
		size /= self.scale
		visible: Set[Chunk] = set()
		for chunks in self.chunks:
			for chunk in chunks:
				if (-256 < chunk.x * 16 + size.width() // 2 - self.cam_x <= size.width() and
					-256 < chunk.y * 16 + size.height() // 2 - self.cam_y <= size.height()):
					visible.add(chunk)
		print(len(visible))
		tx = -self.cam_x + size.width() // 2
		ty = -self.cam_y + size.height() // 2
		for chunk in visible:
			chunk.drawLines(painter, tx, ty)
		for chunk in visible:
			chunk.drawDots(painter, tx, ty, self)
		if cursor is not None and self.getDot(cursor.x(), cursor.y()) == self.dots_manager.empty and \
				0 < cursor.x() < 255 and 0 < cursor.y() < 255:
			painter.setOpacity(.5)
			painter.setBrush(self.dots_manager.color(self.turning_player))
			painter.drawEllipse(cursor.x() * 16 + 5 - self.cam_x + size.width() // 2,
			                    cursor.y() * 16 + 5 - self.cam_y + size.height() // 2, 6, 6)
		painter.restore()

	@staticmethod
	def getAdjacent(x: int, y: int) -> Set[Tuple[int, int]]:
		return Dots.circle(x, y, 1)
	
	def getDot(self, x: int, y: int) -> int:
		if 0 <= x < 256 and 0 <= y < 256:
			return self.chunks[x // 16][y // 16].map[x % 16][y % 16]
		return self.dots_manager.empty

	def getEater(self, x: int, y: int) -> int:
		if 0 <= x < 256 and 0 <= y < 256:
			return self.chunks[x // 16][y // 16].eater[x % 16][y % 16]
		return -1

	def getSurrounding(self, x: int, y: int, every: bool = False) -> Set[Tuple[int, int]]:
		return set([i for i in ((x, y - 1), (x + 1, y - 1),
		                        (x + 1, y), (x + 1, y + 1),
		                        (x, y + 1), (x - 1, y + 1),
		                        (x - 1, y), (x - 1, y - 1))
		            if self.getDot(x, y) == self.getDot(*i) or every])
	
	def _findEmpty(self, x: int, y: int) -> Tuple[int, int]:
		radius = 0
		x //= 16
		y //= 16
		while (radius < x and radius < y and radius < 15 - x and radius < 15 - y and
		       all(not self.chunks[i][j].isEmpty() for i, j in self.circle(x, y, radius))):
			radius += 1
		for i, j in self.circle(x, y, radius):
			if self.chunks[i][j].isEmpty():
				return i * 16 + 8, j * 16 + 8
			if i in {0, 15} or j in {0, 15}:
				return i * 16, j * 16
	
	def findNewTracks(self, x: int, y: int) -> None:
		# This is A* algorithm
		open_tracks: Set[Tuple[Tuple[int, int], ...]] = {((x, y),)}
		used = {(x, y)}
		while open_tracks:
			open_tracks_copy = open_tracks.copy()
			open_tracks.clear()
			for track in open_tracks_copy:
				for pos in self.getSurrounding(*track[-1]):
					if pos not in used:
						open_tracks.add(track + (pos,))
						used.add(pos)
					elif len(track) >= 3 and pos == (x, y):
						self.addTrack(track + (pos,))
					elif pos != (x, y):
						for other in open_tracks_copy.union(open_tracks):
							if other[1] != track[1] and pos == other[-1]:
								self.addTrack(track + other[::-1])
	
	def settings(self) -> DotsSettings:
		return self.dots_manager.settings
	
	def findNewEaten(self, x: int, y: int) -> None:
		empty_dot = self._findEmpty(x, y)
		for xx, yy in self.getAdjacent(x, y):
			if self.getDot(xx, yy) == self.getDot(x, y):
				continue
			
			used: Set[Tuple[int, int]] = set()
			open_pos: Set[Tuple[int, int]] = {(xx, yy)}
			while open_pos and empty_dot not in used:
				open_copy = open_pos.copy()
				open_pos.clear()
				for pos in open_copy:
					for new in self.getAdjacent(*pos):
						if new not in used and self.getDot(*new) != self.getDot(x, y):
							open_pos.add(new)
							if new[0] in {0, 255} or new[1] in {0, 255}:
								break
					used.add(pos)
			if not open_pos:
				for i, j in used:
					if self.dotOwner(i, j) not in {-1, self.dots_manager.player(self.getDot(x, y))}:
						self.counters[self.dots_manager.player(self.getDot(x, y))] += 1
					self.chunks[i // 16][j // 16].map[i % 16][j % 16] = self.dots_manager.eaten(
						self.dots_manager.player(self.getDot(i, j)))
					self.chunks[i // 16][j // 16].eater[i % 16][j % 16] = self.dots_manager.player(self.getDot(x, y))

	def dotOwner(self, x: int, y: int):
		return self.getEater(x, y) if self.getEater(x, y) != -1 else self.dots_manager.player(self.getDot(x, y))

	def translate(self, delta: QPoint, size: QSize) -> None:
		size /= self.scale
		self.cam_x += delta.x()
		self.cam_y += delta.y()
		self.cam_x = max(self.cam_x, size.width() // 2 + self.scale * 16)
		self.cam_x = min(self.cam_x, len(self.chunks) * 256 - size.width() // 2 - self.scale * 16)
		self.cam_y = max(self.cam_y, size.height() // 2 + self.scale * 16)
		self.cam_y = min(self.cam_y, len(self.chunks[0]) * 256 - size.height() // 2 - self.scale * 16)

	def turn(self, x: int, y: int) -> None:
		if self.getDot(x, y) == self.dots_manager.empty and 0 < x < 255 and 0 < y < 255:
			self.chunks[x // 16][y // 16].map[x % 16][y % 16] = self.dots_manager.real(self.turning_player)
			self.findNewTracks(x, y)
			self.findNewEaten(x, y)
			self.turning_player = (self.turning_player + 1) % self.dots_manager.players_number

File: client/dots.cpp
#include "dots.h"
#include <QDebug>
#include <QPainter>
#include <QSize>


Chunk::Chunk(int x, int y)
{
    map = std::vector<std::vector<int>>(16, std::vector<int>(16));

    for (int i = 0; i < 16; i++)
    {
        for (int j = 0; j < 16; j++)
        {
            map[i][j] = 8;
        }
    }

    this->x = x;
    this->y = y;
}

void Chunk::addTrack(QLine track)
{
    if (!tracks.count(QLine(track.p2(), track.p1())))
    {
        tracks.insert(track);
    }
}

std::set<QLine> Chunk::getTracks()
{
    return tracks;
}

int Chunk::dot(int x, int y)
{
    return map[x][y];
}

int Chunk::dot(QPoint pos)
{
    return map[pos.x()][pos.y()];
}

void Chunk::setDot(int x, int y, int dot)
{
    map[x][y] = dot;
}

void Chunk::drawLines(QPainter *painter)
{
    QPen black_pen(Qt::black);
    QPen red_pen(Qt::red);
    black_pen.setWidth(1);
    red_pen.setWidth(1);
    painter->setPen(black_pen);

    for (int i = 0; i < 16; i++)
    {
        painter->drawLine(x * 16, (y + i) * 16 + 8,  // horizontal line
                          x * 16 + 256, (y + i) * 16 + 8);
        painter->drawLine((x + i) * 16 + 8, y * 16,  // vertical line
                          (x + i) * 16 + 8, y * 16 + 256);
    }

    painter->setPen(red_pen);

    if (y == 0)
    {
        painter->drawLine(x * 16, 8, x * 16 + 256, 8);
    }

    if (x == 0)
    {
        painter->drawLine(8, y * 16, 8, y * 16 + 256);
    }

    if (y == 16 * 15)
    {
        painter->drawLine(x * 16, 8 + (y + 15) * 16, x * 16 + 256, 8 + (y + 15) * 16);
    }

    if (x == 16 * 15)
    {
        painter->drawLine(8 + (x + 15) * 16, y * 16, 8 + (x + 15) * 16, y * 16 + 256);
    }
}

void Chunk::drawDots(QPainter *painter, Dots *dots)
{
    for (QLine track : tracks)
    {
        painter->setPen(dsettings()->color(player(dot(track.p1().x() % 16,
                                           track.p1().y() % 16))));
        painter->drawLine(track.p1() * 16 + QPoint(8, 8), track.p2() * 16 + QPoint(8, 8));
    }

    QBrush fill_brush = QBrush(Qt::transparent);
    fill_brush.setStyle(Qt::Dense6Pattern);
    painter->setPen(Qt::transparent);

    for (int x = 0; x < 16; x++)
    {
        for (int y = 0; y < 16; y++)
        {
            painter->setBrush(dsettings()->color(player(dot(x, y))));
            painter->drawEllipse(dotCoord(x, y), 3, 3);

            if (dot(x, y) == 8)
            {
                continue;
            }

            if (dots->isFirst(QPoint(x + this->x, y + this->y)))
            {
                painter->setBrush(Qt::transparent);
                painter->setPen(dsettings()->color(player(dot(x, y))));
                painter->drawEllipse(dotCoord(x, y), 6, 6);
                painter->setPen(Qt::transparent);
            }

            fill_brush.setColor(dsettings()->color(player(dot(x, y))));
            painter->setBrush(fill_brush);
            std::vector<QPoint> circle_ =
            {
                QPoint(x, y - 1),
                QPoint(x + 1, y),
                QPoint(x, y + 1),
                QPoint(x - 1, y)
            };

            for (int i = 0; i < 4; i++)
            {
                if (dots->dot(circle_[i] + QPoint(this->x, this->y)) == dot(x, y) &&
                        dot(x, y) == dots->dot(circle_[(i + 1) % 4] + QPoint(this->x, this->y)))
                {
                    QPoint mono_triang[3] =
                    {
                        dotCoord(x, y),
                        dotCoord(circle_[i]),
                        dotCoord(circle_[(i + 1) % 4])
                    };
                    painter->drawPolygon(mono_triang, 3);
                }
            }

            fill_brush.setColor(dsettings()->color(eater(dot(x, y))));
            painter->setBrush(fill_brush);
            QPoint poly[4] =
            {
                dotCoord(x, y - 1),
                dotCoord(x + 1, y),
                dotCoord(x, y + 1),
                dotCoord(x - 1, y)
            };
            painter->drawPolygon(poly, 4);
            QPoint some_points[4] =
            {
                QPoint(x - 1, y - 1),
                QPoint(x - 1, y + 1),
                QPoint(x + 1, y - 1),
                QPoint(x + 1, y + 1)
            };

            for (QPoint pos : some_points)
            {
                if (player(dots->dot(pos.x() + this->x, pos.y() + this->y)) == eater(dot(x, y)))
                {
                    QPoint triang[3] =
                    {
                        dotCoord(pos),
                        dotCoord(pos.x(), y),
                        dotCoord(x, pos.y())
                    };
                    painter->drawPolygon(triang, 3);
                }
            }
        }
    }
}

QPoint Chunk::dotCoord(int x, int y)
{
    return QPoint((this->x + x) * 16 + 8, (this->y + y) * 16 + 8);
}

QPoint Chunk::dotCoord(QPoint pos)
{
    return dotCoord(pos.x(), pos.y());
}

//#######################################
Dots::Dots()
{
    chunks = {};

    for (int i = 0; i < 16; i++)
    {
        chunks.push_back({});

        for (int j = 0; j < 16; j++)
        {
            chunks[i].push_back(Chunk(i * 16, j * 16));
        }
    }
}

void Dots::addTrack(std::vector<QPoint> track)
{
    QPoint max_dot, min_dot;

    for (unsigned i = 0; i < track.size() - 1; i++)
    {
        max_dot = track[i];
        min_dot = track[i + 1];

        if (max_dot < min_dot)
        {
            std::swap(min_dot, max_dot);
        }

        if (max_dot.x() != min_dot.x() and max_dot.y() != min_dot.y())
        {
            QLine intersec = QLine(max_dot.x(), min_dot.y(), min_dot.x(), max_dot.y());

            if (chunk(intersec.x1(), intersec.y1())->getTracks().count(intersec))
            {
                continue;
            }
        }

        chunk(max_dot.x(), max_dot.y())->addTrack(QLine(min_dot, max_dot));
        chunk(min_dot.x(), min_dot.y())->addTrack(QLine(min_dot, max_dot));
    }
}

std::set<QPoint> Dots::surrounding(int x, int y)
{
    std::set<QPoint> answer;

    for (int dx = -1; dx <= 1; dx++)
    {
        for (int dy = -1; dy <= 1; dy++)
        {
            if (!(dx == 0 and dy == 0) and dot(x + dx, y + dy) == dot(x, y))
            {
                answer.insert(QPoint(x + dx, y + dy));
            }
        }
    }

    return answer;
}

double Dots::getScale()
{
    return scale;
}

QPoint Dots::camPos()
{
    return QPoint(cam_x, cam_y);
}

bool Dots::isFirst(QPoint pos)
{
    return pos == first1 or pos == second1;
}

void Dots::changeScale(double delta, QSize size)
{
    scale += delta;
    scale = std::min(scale, 6.0);
    scale = std::max(scale, 1.5);
    size /= scale;
    normalize(size);
}

Chunk *Dots::chunk(int x, int y)
{
    int cx = x / 16, cy = y / 16;
    return &chunks[cx][cy];
}

int Dots::dot(int x, int y)
{
    return chunk(x, y)->dot(x % 16, y % 16);
}

int Dots::dot(QPoint pos)
{
    return dot(pos.x(), pos.y());
}

void Dots::translate(QPoint delta, QSize size)
{
    size /= scale;
    cam_x += delta.x();
    cam_y += delta.y();
    normalize(size);
}

void Dots::normalize(QSize size)
{
    cam_x = std::max(cam_x, size.width() / 2);
    cam_x = std::min(cam_x, 4096 - size.width() / 2);
    cam_y = std::max(cam_y, size.height() / 2);
    cam_y = std::min(cam_y, 4096 - size.height() / 2);
}

void Dots::findNewTracks(int x, int y)
{
    std::set<std::vector<QPoint>> open_tracks = {{QPoint(x, y)}}, open_tracks_copy;
    std::set<QPoint> used = {QPoint(x, y)};
    std::vector<QPoint> new_track;

    while (!open_tracks.empty())
    {
        open_tracks_copy = open_tracks;
        open_tracks = {};

        for (std::vector<QPoint> track : open_tracks_copy)
        {
            for (QPoint pos : surrounding(track.back().x(), track.back().y()))
            {
                if (!used.count(pos))
                {
                    new_track = track;
                    new_track.push_back(pos);
                    open_tracks.insert(new_track);
                    used.insert(pos);
                }
                else if (track.size() >= 3 and pos == QPoint(x, y))
                {
                    addTrack(track + std::vector<QPoint>(1, pos));
                }
                else if (pos != QPoint(x, y))
                {
                    for (std::vector<QPoint> other : open_tracks_copy)
                    {
                        if (other[1] != track[1] and pos == other.back())
                        {
                            addTrack(track + reversed(other));
                        }
                    }

                    for (std::vector<QPoint> other : open_tracks)
                    {
                        if (other[1] != track[1] and pos == other.back())
                        {
                            addTrack(track + reversed(other));
                        }
                    }
                }
            }
        }
    }
}

void Dots::findNewEaten(int x, int y)
{
    std::set<QPoint> open, open_copy, used;
    bool not_eaten;

    for (QPoint start : circle(x, y))
    {
        used = {};
        open = {start};
        not_eaten = false;

        while (!open.empty() and !not_eaten)
        {
            open_copy = open;
            open = {};

            for (QPoint pos : open_copy)
            {
                if (!QRect(min_x, min_y, max_x - min_x, max_y - min_y).contains(pos))
                {
                    not_eaten = true;
                    break;
                }

                for (QPoint p : circle(pos.x(), pos.y()))
                {
                    if (!used.count(p) and !open.count(p) and dot(p.x(), p.y()) != dot(x, y))
                    {
                        open.insert(p);
                    }
                }

                used.insert(pos);
            }
        }

        if (!not_eaten)
        {
            for (QPoint pos : used)
            {
                chunk(pos.x(), pos.y())->setDot(pos.x() % 16, pos.y() % 16,
                                                player(dot(pos)) + 3 * player(dot(x, y)));
            }
        }
    }
}

void Dots::turn(QPoint pos)
{
    if (winner() != 2)
    {
        return;
    }

    if (QRect(1, 1, 255, 255).contains(pos.x(), pos.y()) and dot(pos.x(), pos.y()) == 8)
    {
        chunk(pos.x(), pos.y())->setDot(pos.x() % 16, pos.y() % 16, turning_player + 6);
        max_x = std::max(max_x, pos.x());
        max_y = std::max(max_y, pos.y());
        min_x = std::min(min_x, pos.x());
        min_y = std::min(min_y, pos.y());
        findNewTracks(pos.x(), pos.y());
        findNewEaten(pos.x(), pos.y());

        if (first1 == QPoint(-1, -1) and turning_player == 0)
        {
            first1 = pos;
        }

        if (second1 == QPoint(-1, -1) and turning_player == 1)
        {
            second1 = pos;
        }

        turning_player = (turning_player + 1) % 2;
    }
}

int Dots::winner()
{
    if (first1 != QPoint(-1, -1) and eater(dot(first1)) == 1)
    {
        return 1;
    }

    if (second1 != QPoint(-1, -1) and eater(dot(second1)) == 0)
    {
        return 0;
    }

    return 2;
}

void Dots::draw(QSize size, QPainter *painter, QPoint cursor)
{
    painter->save();
    painter->scale(scale, scale);
    size /= scale;
    normalize(size);
    int tx = size.width() / 2 - cam_x, ty = size.height() / 2 - cam_y;
    painter->translate(tx, ty);
    std::set<Chunk *> visible = {};

    for (int x = 0; x < 256; x += 16)
    {
        for (int y = 0; y < 256; y += 16)
        {
            if (-256 < x * 16 + tx and x * 16 + tx <= size.width() and
                    - 256 < y * 16 + ty and y * 16 + ty <= size.height())
            {
                visible.insert(chunk(x, y));
                chunk(x, y)->drawLines(painter);
            }
        }
    }

    for (Chunk *chunk : visible)
    {
        chunk->drawDots(painter, this);
    }

    painter->setPen(Qt::transparent);

    if (QRect(1, 1, 255, 255).contains(cursor) and
            dot(cursor.x(), cursor.y()) == 8)
    {
        painter->setBrush(dsettings()->color(turning_player));
        painter->setOpacity(0.5);
        painter->drawEllipse(cursor * 16 + QPoint(8, 8), 3, 3);

        if (first1 == QPoint(-1, -1) or second1 == QPoint(-1, -1))
        {
            painter->setBrush(Qt::transparent);
            painter->setPen(dsettings()->color(turning_player));
            painter->drawEllipse(cursor * 16 + QPoint(8, 8), 6, 6);
        }
    }

    painter->restore();
}
